File: src/common/tarjan/__tests__/tarjan_test.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open OUnit2

module StringNode = struct
  type t = string

  let compare = String.compare

  let to_string x = x
end

module StringMap = struct
  type key = string

  type value = SSet.t

  type t = SSet.t SMap.t

  let find k m = SMap.find k m
end

module StringTarjan = Tarjan.Make (StringNode) (SSet) (StringMap)

(* Helper: convert list of edges to a graph map *)
let graph_of_list edges =
  List.fold_left (fun acc (k, vs) -> SMap.add k (SSet.of_list vs) acc) SMap.empty edges

(* Helper: create a set from a list *)
let roots_of_list lst = SSet.of_list lst

(* Helper: normalize an SCC (sort elements within each component) *)
let normalize_scc (hd, tl) =
  let all = hd :: tl in
  let sorted = List.sort String.compare all in
  match sorted with
  | [] -> failwith "impossible: empty SCC"
  | x :: xs -> (x, xs)

(* Helper: compare two SCCs for equality (order within component doesn't matter) *)
let scc_equal a b =
  let a_norm = normalize_scc a in
  let b_norm = normalize_scc b in
  a_norm = b_norm

(* Helper: compare two lists of SCCs *)
let sccs_equal expected actual =
  List.length expected = List.length actual && List.for_all2 scc_equal expected actual

(* Helper: pretty print an SCC *)
let string_of_scc (hd, tl) =
  let all = hd :: tl in
  Printf.sprintf "[%s]" (String.concat ", " all)

(* Helper: pretty print a list of SCCs *)
let string_of_sccs sccs = Printf.sprintf "[%s]" (String.concat "; " (List.map string_of_scc sccs))

(* Helper: assert SCCs equal with nice error messages *)
let assert_sccs_equal ~ctxt expected actual =
  assert_equal ~ctxt ~cmp:sccs_equal ~printer:string_of_sccs expected actual

let tests =
  "tarjan"
  >::: [
         (* Basic Cases *)
         ( "empty_roots" >:: fun ctxt ->
           let graph = graph_of_list [] in
           let roots = roots_of_list [] in
           let result = StringTarjan.topsort ~roots graph in
           assert_sccs_equal ~ctxt [] result
         );
         ( "single_node_no_edges" >:: fun ctxt ->
           let graph = graph_of_list [("a", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           assert_sccs_equal ~ctxt [("a", [])] result
         );
         ( "single_node_self_loop" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["a"])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           assert_sccs_equal ~ctxt [("a", [])] result
         );
         ( "two_nodes_no_edges" >:: fun ctxt ->
           let graph = graph_of_list [("a", []); ("b", [])] in
           let roots = roots_of_list ["a"; "b"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Order may vary, so check both possibilities *)
           let result_normalized = List.map normalize_scc result |> List.sort compare in
           let expected = [("a", []); ("b", [])] |> List.map normalize_scc |> List.sort compare in
           assert_equal ~ctxt expected result_normalized
         );
         ( "two_nodes_one_edge" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"]); ("b", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Dependent first order: [A], [B] *)
           assert_sccs_equal ~ctxt [("a", []); ("b", [])] result
         );
         ( "two_nodes_cycle" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"]); ("b", ["a"])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Single SCC containing both nodes *)
           assert_equal ~ctxt 1 (List.length result);
           let (hd, tl) = List.hd result in
           let all_nodes = hd :: tl |> List.sort String.compare in
           assert_equal ~ctxt ["a"; "b"] all_nodes
         );
         (* Chain/Linear Cases *)
         ( "linear_chain_3_nodes" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Dependent first order: A, B, C *)
           assert_sccs_equal ~ctxt [("a", []); ("b", []); ("c", [])] result
         );
         ( "linear_chain_5_nodes" >:: fun ctxt ->
           let graph =
             graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", ["d"]); ("d", ["e"]); ("e", [])]
           in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Dependent first order: A, B, C, D, E *)
           assert_sccs_equal ~ctxt [("a", []); ("b", []); ("c", []); ("d", []); ("e", [])] result
         );
         (* Cycle Cases *)
         ( "triangle_cycle" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", ["a"])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Single SCC containing all three nodes *)
           assert_equal ~ctxt 1 (List.length result);
           let (hd, tl) = List.hd result in
           let all_nodes = hd :: tl |> List.sort String.compare in
           assert_equal ~ctxt ["a"; "b"; "c"] all_nodes
         );
         ( "cycle_with_tail" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", ["b"])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Two SCCs: {A} and {B, C} *)
           assert_equal ~ctxt 2 (List.length result);
           (* First SCC should be {A} *)
           let (hd1, tl1) = List.nth result 0 in
           assert_equal ~ctxt "a" hd1;
           assert_equal ~ctxt [] tl1;
           (* Second SCC should be {B, C} *)
           let (hd2, tl2) = List.nth result 1 in
           let scc2 = hd2 :: tl2 |> List.sort String.compare in
           assert_equal ~ctxt ["b"; "c"] scc2
         );
         ( "two_separate_cycles" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"]); ("b", ["a"]); ("c", ["d"]); ("d", ["c"])] in
           let roots = roots_of_list ["a"; "c"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Two SCCs: {A, B} and {C, D} *)
           assert_equal ~ctxt 2 (List.length result);
           (* Check both SCCs, order may vary *)
           let sccs =
             result
             |> List.map (fun (hd, tl) -> hd :: tl |> List.sort String.compare)
             |> List.sort compare
           in
           assert_equal ~ctxt [["a"; "b"]; ["c"; "d"]] sccs
         );
         (* Complex Graph Patterns *)
         ( "diamond_pattern" >:: fun ctxt ->
           let graph = graph_of_list [("a", ["b"; "c"]); ("b", ["d"]); ("c", ["d"]); ("d", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Four SCCs: A, then B and C (order may vary), then D *)
           assert_equal ~ctxt 4 (List.length result);
           (* A should be first (no dependents) *)
           let (hd0, tl0) = List.nth result 0 in
           assert_equal ~ctxt "a" hd0;
           assert_equal ~ctxt [] tl0;
           (* D should be last (no dependencies) *)
           let (hd3, tl3) = List.nth result 3 in
           assert_equal ~ctxt "d" hd3;
           assert_equal ~ctxt [] tl3
         );
         ( "figure_8" >:: fun ctxt ->
           (* A→B→C, B→A (first cycle), C→D→E, D→C (second cycle) *)
           let graph =
             graph_of_list [("a", ["b"]); ("b", ["a"; "c"]); ("c", ["d"]); ("d", ["c"])]
           in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Two SCCs: {A, B} and {C, D} *)
           assert_equal ~ctxt 2 (List.length result);
           let sccs = result |> List.map (fun (hd, tl) -> hd :: tl |> List.sort String.compare) in
           (* {A, B} should come before {C, D} because {A,B} depends on {C,D} *)
           assert_equal ~ctxt ["a"; "b"] (List.nth sccs 0);
           assert_equal ~ctxt ["c"; "d"] (List.nth sccs 1)
         );
         ( "nested_cycles" >:: fun ctxt ->
           (* Outer: A→B→C→A, Inner: B→D→B *)
           let graph =
             graph_of_list [("a", ["b"]); ("b", ["c"; "d"]); ("c", ["a"]); ("d", ["b"])]
           in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* All four nodes in one SCC *)
           assert_equal ~ctxt 1 (List.length result);
           let (hd, tl) = List.hd result in
           let all_nodes = hd :: tl |> List.sort String.compare in
           assert_equal ~ctxt ["a"; "b"; "c"; "d"] all_nodes
         );
         ( "multiple_sccs_varying_sizes" >:: fun ctxt ->
           (* {A} alone, {B, C} cycle, {D, E, F} cycle *)
           let graph =
             graph_of_list
               [
                 ("a", ["b"]);
                 ("b", ["c"]);
                 ("c", ["b"; "d"]);
                 ("d", ["e"]);
                 ("e", ["f"]);
                 ("f", ["d"]);
               ]
           in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Three SCCs *)
           assert_equal ~ctxt 3 (List.length result);
           let sccs = result |> List.map (fun (hd, tl) -> hd :: tl |> List.sort String.compare) in
           (* Check SCC sizes *)
           let scc_sizes = List.map List.length sccs |> List.sort compare in
           assert_equal ~ctxt [1; 2; 3] scc_sizes
         );
         ( "large_scc" >:: fun ctxt ->
           (* Six nodes all in one big cycle: A→B→C→D→E→F→A *)
           let graph =
             graph_of_list
               [("a", ["b"]); ("b", ["c"]); ("c", ["d"]); ("d", ["e"]); ("e", ["f"]); ("f", ["a"])]
           in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Single SCC with 6 nodes *)
           assert_equal ~ctxt 1 (List.length result);
           let (hd, tl) = List.hd result in
           let all_nodes = hd :: tl |> List.sort String.compare in
           assert_equal ~ctxt ["a"; "b"; "c"; "d"; "e"; "f"] all_nodes
         );
         (* Edge Cases *)
         ( "partial_reachability" >:: fun ctxt ->
           (* Graph has A→B, C→D, but only A is a root *)
           let graph = graph_of_list [("a", ["b"]); ("b", []); ("c", ["d"]); ("d", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Only A and B should be visited *)
           assert_sccs_equal ~ctxt [("a", []); ("b", [])] result
         );
         ( "subset_of_roots" >:: fun ctxt ->
           (* A→B→C, only A is a root, B and C discovered via edges *)
           let graph = graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           assert_sccs_equal ~ctxt [("a", []); ("b", []); ("c", [])] result
         );
         ( "all_nodes_as_roots" >:: fun ctxt ->
           (* A→B, B→C, but all three are roots *)
           let graph = graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", [])] in
           let roots = roots_of_list ["a"; "b"; "c"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Same result as if only A was root *)
           assert_sccs_equal ~ctxt [("a", []); ("b", []); ("c", [])] result
         );
         ( "star_pattern" >:: fun ctxt ->
           (* Hub A points to B, C, D, E (no back edges) *)
           let graph =
             graph_of_list [("a", ["b"; "c"; "d"; "e"]); ("b", []); ("c", []); ("d", []); ("e", [])]
           in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Five SCCs, all singleton, A should be first *)
           assert_equal ~ctxt 5 (List.length result);
           let (hd0, tl0) = List.nth result 0 in
           assert_equal ~ctxt "a" hd0;
           assert_equal ~ctxt [] tl0;
           (* Last four should be B, C, D, E in some order *)
           let last_four =
             List.filteri (fun i _ -> i > 0) result
             |> List.map (fun (hd, _) -> hd)
             |> List.sort String.compare
           in
           assert_equal ~ctxt ["b"; "c"; "d"; "e"] last_four
         );
         ( "reverse_star" >:: fun ctxt ->
           (* B, C, D, E all point to hub A *)
           let graph =
             graph_of_list [("b", ["a"]); ("c", ["a"]); ("d", ["a"]); ("e", ["a"]); ("a", [])]
           in
           let roots = roots_of_list ["b"; "c"; "d"; "e"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Five SCCs, A should appear somewhere (exact position depends on visit order) *)
           assert_equal ~ctxt 5 (List.length result);
           (* Verify A is present as a singleton SCC *)
           let has_a_singleton = result |> List.exists (fun (hd, tl) -> hd = "a" && tl = []) in
           assert_bool "A should be a singleton SCC" has_a_singleton;
           (* Verify all nodes are present *)
           let all_nodes = result |> List.map (fun (hd, _) -> hd) |> List.sort String.compare in
           assert_equal ~ctxt ["a"; "b"; "c"; "d"; "e"] all_nodes
         );
         (* Topological Order Verification *)
         ( "verify_reverse_topo_order" >:: fun _ctxt ->
           (* A→B→C→D, verify dependents appear before dependencies *)
           let graph = graph_of_list [("a", ["b"]); ("b", ["c"]); ("c", ["d"]); ("d", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Build index map: node -> position in result *)
           let indices =
             result |> List.mapi (fun i (hd, _) -> (hd, i)) |> List.to_seq |> Hashtbl.of_seq
           in
           (* For each edge u→v in graph, verify index[u] < index[v] (dependent before dependency) *)
           let check_edge u v =
             let u_idx = Hashtbl.find indices u in
             let v_idx = Hashtbl.find indices v in
             assert_bool
               (Printf.sprintf "Edge %s→%s: expected %d < %d" u v u_idx v_idx)
               (u_idx < v_idx)
           in
           check_edge "a" "b";
           check_edge "b" "c";
           check_edge "c" "d"
         );
         ( "cross_edges_dont_create_sccs" >:: fun ctxt ->
           (* A→B→C with cross-edge A→C (doesn't create cycle) *)
           let graph = graph_of_list [("a", ["b"; "c"]); ("b", ["c"]); ("c", [])] in
           let roots = roots_of_list ["a"] in
           let result = StringTarjan.topsort ~roots graph in
           (* Three separate SCCs (cross-edge doesn't form cycle) *)
           assert_sccs_equal ~ctxt [("a", []); ("b", []); ("c", [])] result
         );
       ]

let () = run_test_tt_main tests

File: rust_port/crates/flow_common_tarjan/src/tarjan_test.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::collections::BTreeSet;
use std::collections::HashMap;
use std::sync::Arc;

use dupe::Dupe;
use flow_common_utils::graph::Graph;
use vec1::Vec1;

use crate::topsort;

// Helper to create a graph from key-value pairs
fn graph_from_pairs(pairs: &[(&str, &[&str])]) -> Graph<Arc<str>> {
    let map: BTreeMap<Arc<str>, BTreeSet<Arc<str>>> = pairs
        .iter()
        .map(|(key, values)| (Arc::from(*key), set_from_slice(values)))
        .collect();
    Graph::of_map(map)
}

// Helper to create a set from a slice
fn set_from_slice(items: &[&str]) -> BTreeSet<Arc<str>> {
    items.iter().map(|s| Arc::from(*s)).collect()
}

// Helper to normalize a component (sort elements)
fn normalize_component(comp: &Vec1<Arc<str>>) -> Vec<String> {
    let mut items: Vec<String> = comp.iter().map(|s| s.to_string()).collect();
    items.sort();
    items
}

// Helper to compare two components for equality (order within doesn't matter)
fn components_equal(a: &Vec1<Arc<str>>, b: &Vec1<Arc<str>>) -> bool {
    normalize_component(a) == normalize_component(b)
}

// Helper to compare two SCC results
fn results_equal(expected: &[Vec1<Arc<str>>], actual: &[Vec1<Arc<str>>]) -> bool {
    if expected.len() != actual.len() {
        return false;
    }
    expected
        .iter()
        .zip(actual.iter())
        .all(|(e, a)| components_equal(e, a))
}

// Helper to create Vec1 from a slice
fn vec1_from_slice(items: &[&str]) -> Vec1<Arc<str>> {
    let v: Vec<Arc<str>> = items.iter().map(|s| Arc::from(*s)).collect();
    Vec1::try_from_vec(v).expect("Expected non-empty vec")
}

#[test]
fn test_empty_roots() {
    let graph = graph_from_pairs(&[]);
    let roots = set_from_slice(&[]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    assert_eq!(0, result.len());
}

#[test]
fn test_single_node_no_edges() {
    let graph = graph_from_pairs(&[("a", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    let expected = vec![vec1_from_slice(&["a"])];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_single_node_self_loop() {
    let graph = graph_from_pairs(&[("a", &["a"])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    let expected = vec![vec1_from_slice(&["a"])];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_two_nodes_no_edges() {
    let graph = graph_from_pairs(&[("a", &[]), ("b", &[])]);
    let roots = set_from_slice(&["a", "b"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Order may vary, check both are singleton SCCs
    assert_eq!(2, result.len());
    let mut nodes: Vec<String> = result
        .iter()
        .flat_map(|comp| comp.iter().map(|s| s.to_string()))
        .collect();
    nodes.sort();
    assert_eq!(vec!["a", "b"], nodes);
}

#[test]
fn test_two_nodes_one_edge() {
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Dependent-first order: A depends on B, so [A], [B]
    let expected = vec![vec1_from_slice(&["a"]), vec1_from_slice(&["b"])];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_two_nodes_cycle() {
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["a"])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Single SCC containing both nodes
    assert_eq!(1, result.len());
    let nodes = normalize_component(&result[0]);
    assert_eq!(vec!["a", "b"], nodes);
}

#[test]
fn test_linear_chain_3_nodes() {
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["c"]), ("c", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Dependent-first order: A→B→C means [A], [B], [C]
    let expected = vec![
        vec1_from_slice(&["a"]),
        vec1_from_slice(&["b"]),
        vec1_from_slice(&["c"]),
    ];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_linear_chain_5_nodes() {
    let graph = graph_from_pairs(&[
        ("a", &["b"]),
        ("b", &["c"]),
        ("c", &["d"]),
        ("d", &["e"]),
        ("e", &[]),
    ]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Dependent-first order: A→B→C→D→E means [A], [B], [C], [D], [E]
    let expected = vec![
        vec1_from_slice(&["a"]),
        vec1_from_slice(&["b"]),
        vec1_from_slice(&["c"]),
        vec1_from_slice(&["d"]),
        vec1_from_slice(&["e"]),
    ];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_triangle_cycle() {
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["c"]), ("c", &["a"])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Single SCC containing all three nodes
    assert_eq!(1, result.len());
    let nodes = normalize_component(&result[0]);
    assert_eq!(vec!["a", "b", "c"], nodes);
}

#[test]
fn test_cycle_with_tail() {
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["c"]), ("c", &["b"])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Two SCCs: A depends on {B,C}, so [A], [{B,C}]
    assert_eq!(2, result.len());
    // First SCC should be {A}
    let expected_a = vec1_from_slice(&["a"]);
    assert!(components_equal(&expected_a, &result[0]));
    // Second SCC should be {B, C}
    let scc2 = normalize_component(&result[1]);
    assert_eq!(vec!["b", "c"], scc2);
}

#[test]
fn test_two_separate_cycles() {
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["a"]), ("c", &["d"]), ("d", &["c"])]);
    let roots = set_from_slice(&["a", "c"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Two SCCs: {A, B} and {C, D}
    assert_eq!(2, result.len());
    let mut sccs: Vec<Vec<String>> = result.iter().map(normalize_component).collect();
    sccs.sort();
    assert_eq!(vec![vec!["a", "b"], vec!["c", "d"]], sccs);
}

#[test]
fn test_diamond_pattern() {
    let graph = graph_from_pairs(&[("a", &["b", "c"]), ("b", &["d"]), ("c", &["d"]), ("d", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Four SCCs: A first (depends on all), then B and C (order may vary), then D last
    assert_eq!(4, result.len());
    // A should be first (depends on all others)
    let expected_a = vec1_from_slice(&["a"]);
    assert!(components_equal(&expected_a, &result[0]));
    // D should be last (no dependencies)
    let expected_d = vec1_from_slice(&["d"]);
    assert!(components_equal(&expected_d, &result[3]));
}

#[test]
fn test_figure_8() {
    // A→B→C, B→A (first cycle), C→D→E, D→C (second cycle)
    let graph = graph_from_pairs(&[
        ("a", &["b"]),
        ("b", &["a", "c"]),
        ("c", &["d"]),
        ("d", &["c"]),
    ]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Two SCCs: {A, B} depends on {C, D}, so [{A,B}], [{C,D}]
    assert_eq!(2, result.len());
    let sccs: Vec<Vec<String>> = result.iter().map(normalize_component).collect();
    // {A, B} should come before {C, D} because {A,B}→{C,D}
    assert_eq!(vec!["a", "b"], sccs[0]);
    assert_eq!(vec!["c", "d"], sccs[1]);
}

#[test]
fn test_nested_cycles() {
    // Outer: A→B→C→A, Inner: B→D→B
    let graph = graph_from_pairs(&[
        ("a", &["b"]),
        ("b", &["c", "d"]),
        ("c", &["a"]),
        ("d", &["b"]),
    ]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // All four nodes in one SCC
    assert_eq!(1, result.len());
    let nodes = normalize_component(&result[0]);
    assert_eq!(vec!["a", "b", "c", "d"], nodes);
}

#[test]
fn test_multiple_sccs_varying_sizes() {
    // {A} alone, {B, C} cycle, {D, E, F} cycle
    let graph = graph_from_pairs(&[
        ("a", &["b"]),
        ("b", &["c"]),
        ("c", &["b", "d"]),
        ("d", &["e"]),
        ("e", &["f"]),
        ("f", &["d"]),
    ]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Three SCCs
    assert_eq!(3, result.len());
    let mut scc_sizes: Vec<usize> = result.iter().map(|comp| comp.len()).collect();
    scc_sizes.sort();
    assert_eq!(vec![1, 2, 3], scc_sizes);
}

#[test]
fn test_large_scc() {
    // Six nodes all in one big cycle: A→B→C→D→E→F→A
    let graph = graph_from_pairs(&[
        ("a", &["b"]),
        ("b", &["c"]),
        ("c", &["d"]),
        ("d", &["e"]),
        ("e", &["f"]),
        ("f", &["a"]),
    ]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Single SCC with 6 nodes
    assert_eq!(1, result.len());
    let nodes = normalize_component(&result[0]);
    assert_eq!(vec!["a", "b", "c", "d", "e", "f"], nodes);
}

#[test]
fn test_partial_reachability() {
    // Graph has A→B, C→D, but only A is a root
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &[]), ("c", &["d"]), ("d", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Only A and B should be visited, dependent-first: [A], [B]
    let expected = vec![vec1_from_slice(&["a"]), vec1_from_slice(&["b"])];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_subset_of_roots() {
    // A→B→C, only A is a root, B and C discovered via edges
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["c"]), ("c", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Dependent-first: [A], [B], [C]
    let expected = vec![
        vec1_from_slice(&["a"]),
        vec1_from_slice(&["b"]),
        vec1_from_slice(&["c"]),
    ];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_all_nodes_as_roots() {
    // A→B, B→C, but all three are roots
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["c"]), ("c", &[])]);
    let roots = set_from_slice(&["a", "b", "c"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Same result as if only A was root: [A], [B], [C]
    let expected = vec![
        vec1_from_slice(&["a"]),
        vec1_from_slice(&["b"]),
        vec1_from_slice(&["c"]),
    ];
    assert!(results_equal(&expected, &result));
}

#[test]
fn test_star_pattern() {
    // Hub A points to B, C, D, E (no back edges)
    let graph = graph_from_pairs(&[
        ("a", &["b", "c", "d", "e"]),
        ("b", &[]),
        ("c", &[]),
        ("d", &[]),
        ("e", &[]),
    ]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Five SCCs, all singleton, A should be first (depends on all)
    assert_eq!(5, result.len());
    let expected_a = vec1_from_slice(&["a"]);
    assert!(components_equal(&expected_a, &result[0]));
    // Last four should be B, C, D, E in some order
    let mut last_four: Vec<String> = result[1..5]
        .iter()
        .flat_map(|comp| comp.iter().map(|s| s.to_string()))
        .collect();
    last_four.sort();
    assert_eq!(vec!["b", "c", "d", "e"], last_four);
}

#[test]
fn test_reverse_star() {
    // B, C, D, E all point to hub A
    let graph = graph_from_pairs(&[
        ("b", &["a"]),
        ("c", &["a"]),
        ("d", &["a"]),
        ("e", &["a"]),
        ("a", &[]),
    ]);
    let roots = set_from_slice(&["b", "c", "d", "e"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Five SCCs, {B,C,D,E} should be first (depend on A), A should be last
    assert_eq!(5, result.len());
    let expected_a = vec1_from_slice(&["a"]);
    assert!(components_equal(&expected_a, &result[4]));
}

#[test]
fn test_verify_reverse_topo_order() {
    // A→B→C→D, verify dependents appear before dependencies
    let graph = graph_from_pairs(&[("a", &["b"]), ("b", &["c"]), ("c", &["d"]), ("d", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Build index map: node -> position in result
    let mut indices: HashMap<Arc<str>, usize> = HashMap::new();
    for (i, comp) in result.iter().enumerate() {
        for node in comp.iter() {
            indices.insert(node.dupe(), i);
        }
    }
    // For each edge u→v in graph, verify index[u] < index[v] (dependent before dependency)
    let check_edge = |u: &str, v: &str| {
        let u_idx = indices.get(&Arc::from(u)).unwrap();
        let v_idx = indices.get(&Arc::from(v)).unwrap();
        assert!(
            u_idx < v_idx,
            "Edge {}→{}: expected {} < {}",
            u,
            v,
            u_idx,
            v_idx
        );
    };
    check_edge("a", "b");
    check_edge("b", "c");
    check_edge("c", "d");
}

#[test]
fn test_cross_edges_dont_create_sccs() {
    // A→B→C with cross-edge A→C (doesn't create cycle)
    let graph = graph_from_pairs(&[("a", &["b", "c"]), ("b", &["c"]), ("c", &[])]);
    let roots = set_from_slice(&["a"]);
    let result = topsort(roots.iter().map(|r| r.dupe()), &graph);
    // Three separate SCCs (cross-edge doesn't form cycle): [A], [B], [C]
    let expected = vec![
        vec1_from_slice(&["a"]),
        vec1_from_slice(&["b"]),
        vec1_from_slice(&["c"]),
    ];
    assert!(results_equal(&expected, &result));
}
